README.md
# Manta Front End

Web app for interacting with Manta Network and Calamari Network, and Dolphin Testnet

## Getting started (for opensource development)

1. Clone the repo.

   `git clone git@github.com:Manta-Network/manta-front-end.git`
2. Change to the repo directory.

   `cd manta-front-end`

3. Install dependencies and start the dev server, waiting for hmr working.

   `yarn && yarn start`

4. Clone manta-signer repo.

   `git clone git@github.com:Manta-Network/manta-signer.git`

5. Change to the tauri-ui directory.

   `cd manta-signer/ui/src-tauri`

6. Install tauri and build manta-signer locally.

    ```bash
    cargo install tauri-cli

    cargo tauri build --features=unsafe-disable-cors
    ```

7. Open the local disable-cors version of manta-signer by following [guides](https://docs.manta.network/docs/guides/MantaSigner).

8. Follow [guides](https://docs.manta.network/docs/guides/DolphinPay) and try MantaPay on Dolphin Testnet.

9. Happy coding!

## Versioning

The first digit of the version (x in x.y.z) refers to exceptional massive changes across manta front end, other full stack repositories, and the runtime, such as the upgrade to testnet V2 (reusable addresses, major cryptographic changes).

The second digit of the version (y in x.y.z) refers to other changes that are coordinated across multiple full stack repositories, and perhaps also the runtime. For example, an upgraded `manta-wasm-wallet` syncing protocol requiring changes to `manta-front-end`, `manta-wasm-wallet` in sdk, and the runtime would fall under this category.

The third digit of the version (z in x.y.z) refers to changes that only affect the front end. A UI restyle would fall under this category.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.1.0] - 2022-06-16

### Added

- [\#145](https://github.com/Manta-Network/manta-front-end/pull/145) Distinguish longer-lasting initial sync from subsequent syncs in UI text
- [\#145](https://github.com/Manta-Network/manta-front-end/pull/145) Add loading animation to '...' text while private balances are syncing to chain
- [\#138](https://github.com/Manta-Network/manta-front-end/pull/138) / [\#160](https://github.com/Manta-Network/manta-front-end/pull/160) Support end-to-end testing by adding metadata to front end components
- [\#126](https://github.com/Manta-Network/manta-front-end/pull/126) Add bug report link to sidebar
- [\#154](https://github.com/Manta-Network/manta-front-end/pull/154) Add versioning rules to readme
- [\#164](https://github.com/Manta-Network/manta-front-end/pull/164) Add downtime modal
- [\#134](https://github.com/Manta-Network/manta-front-end/pull/134) Persist and load privacy toggle / asset selection

### Changed

- [\#141](https://github.com/Manta-Network/manta-front-end/pull/141) Sync to chain more quickly and efficiently using new RPC

### Fixed

- [\#142](https://github.com/Manta-Network/manta-front-end/pull/142) Prevent stale balances from displaying after a transaction while waiting for wallet synchronization
- [\#152](https://github.com/Manta-Network/manta-front-end/pull/152) Add staging branch to CI/CD pipeline

## [2.2.0] - 2022-07-6

### Fixed

- [\#187](https://github.com/Manta-Network/manta-front-end/pull/187) Updates the front end to use the new pull RPC interface, which makes the front end compatible with the current runtime, and reduces wallet sync times
- [\#187](https://github.com/Manta-Network/manta-front-end/pull/187) Fixes a bug such that dolphin public balances won't load

## [2.3.0] - 2022-07-8

### Changed

- [\#184](https://github.com/Manta-Network/manta-front-end/pull/184) Moved inline SVGs to their own files
- [\#176](https://github.com/Manta-Network/manta-front-end/pull/176) Vertically and horizontally centered page content

### Fixed

- [\#183](https://github.com/Manta-Network/manta-front-end/pull/183) Fixed Send Max Amount warning for DOL, which was stretching out the UI for 2 or 3 pixels only on firefox
- [\#183](https://github.com/Manta-Network/manta-front-end/pull/183) Fixed Dolphin icon missing piece
- [\#188](https://github.com/Manta-Network/manta-front-end/pull/188) Fixed sidebar menu disappearing on smaller screens
- [\#182](https://github.com/Manta-Network/manta-front-end/pull/182) Fixed autocomplete getting discolored under some conditions

## [3.0.0] - 2022-10-5

### Added

- [\#252](https://github.com/Manta-Network/manta-front-end/pull/252) Added Calamari staking page and support for multiple networks

## [3.0.1] - 2022-10-11

### Added

- [\#270](https://github.com/Manta-Network/manta-front-end/pull/270) Display previous round rewards and APY estimates adjusted for collator performance on the staking page

## [3.0.2] - 2022-10-12

### Fixed

- [\#272](https://github.com/Manta-Network/manta-front-end/pull/272) Prevent error when user is staked to a node that left the set of collator candidates

## [3.0.3] - 2022-10-19

### Fixed

- [\#275](https://github.com/Manta-Network/manta-front-end/pull/275) Fix display issue on widescreens; fix failure to connect to Talisman wallet

## [3.0.4] - 2022-10-22

### Fixed

- [\#281](https://github.com/Manta-Network/manta-front-end/pull/281) Prevent excess api calls when initializing wallet

## [3.1.0] - 2022-10-28

### Added

- [\#249](https://github.com/Manta-Network/manta-front-end/pull/249) Add FAQ links to missing required software modal
- [\#271](https://github.com/Manta-Network/manta-front-end/pull/271) Display user balance in staking modal; add APY estimates to staking table; add link to collator onboarding docs to staking UI
- [\#273](https://github.com/Manta-Network/manta-front-end/pull/273) Add secondary menu containing social and documentation links + theme toggle

### Changed

- [\#208](https://github.com/Manta-Network/manta-front-end/pull/208) Display transaction success modal when transaction is in block instead of finalized
- [\#210](https://github.com/Manta-Network/manta-front-end/pull/210) Add ids for e2e automation
- [\#235](https://github.com/Manta-Network/manta-front-end/pull/235) Change text displayed while node is syncing
- [\#271](https://github.com/Manta-Network/manta-front-end/pull/271) Show only active collators by default on staking page; reduce staking page width to improve UX on smaller screens

### Removed

- [\#227](https://github.com/Manta-Network/manta-front-end/pull/227) Remove gh-pages dependency target

### Fixed

- [\#228](https://github.com/Manta-Network/manta-front-end/pull/228) Prevent loader and initial sync text from displaying if polkadot.js api becomes disconnected
- [\#259](https://github.com/Manta-Network/manta-front-end/pull/259) Fix typescript type annotation compilation for files with `.ts` extension
- [\#271](https://github.com/Manta-Network/manta-front-end/pull/271) Fix issue with URL parameters being ignored before '#'

### Security

- [\#198](https://github.com/Manta-Network/manta-front-end/pull/198) Bump react-router-dom from 5.3.3 to 6.3.0 (required refactoring of breaking API changes)
- [\#201](https://github.com/Manta-Network/manta-front-end/pull/201) Bump @craco/craco from 6.4.3 to 6.4.5
- [\#117](https://github.com/Manta-Network/manta-front-end/pull/117) Bump node-sass from 6.0.1 to 7.0.1

## [3.1.1] - 2022-12-02

### Added

- [\#310](https://github.com/Manta-Network/manta-front-end/pull/310) Added name for a collator

## [3.1.2] - 2022-12-15

### Changed

- [\#331](https://github.com/Manta-Network/manta-front-end/pull/331) Added or changed names for several collators


src/types/AssetType.ts
// @ts-nocheck
import NETWORK from 'constants/NetworkConstants';
import BN from 'bn.js';

const CalamariAssetIds = {
  KMA: 1,
  KAR: 8,
  AUSD: 9,
  LKSM: 10,
  MOVR: 11,
  KSM: 12,
  USDT: 14,
  DAI: 15,
  USDC: 16,
  ARB: 17,
  LDO: 18,
  SHIB: 19,
  MATIC: 20,
  BNB: 21,
  UNI: 22,
  BUSD: 23,
  LINK: 24,
  APE: 25,
  WBTC: 26,
  WETH: 27
};

const MantaAssetIds = {
  MANTA: 1,
  DOT: 8,
  USDT: 9,
  GLMR: 10,
  ACA: 11,
  LDOT: 12,
  APE: 19,
  USDT: 22,
  DAI: 23,
  USDC: 24,
  WBTC: 26,
  WETH: 27
};

const getAssetIds = (config) => {
  if (config.NETWORK_NAME === NETWORK.CALAMARI) {
    return CalamariAssetIds;
  } else {
    return MantaAssetIds;
  }
};

export default class AssetType {
  assetId: number;
  baseName: string;
  name: string;
  baseTicker: string;
  ticker: string;
  logicalTicker: string;
  icon: string;
  numberOfDecimals: number;
  existentialDeposit: BN;
  existentialDeposit: BN;
  isPrivate: boolean;
  isTestnet: boolean;
  isNativeToken: boolean;
  coingeckoId: string;
  displayDecimals: number;

  constructor(
    assetId,
    baseName,
    baseTicker,
    icon,
    numberOfDecimals,
    existentialDeposit,
    isPrivate,
    coingeckoId,
    isTestnet,
    isNativeToken = false,
    logicalTicker = null,
    displayDecimals = 2
  ) {
    this.assetId = assetId;
    this.baseName = baseName;
    this.baseTicker = baseTicker;
    this.logicalTicker = logicalTicker || baseTicker;
    this.name = AssetType._getFullName(baseName, isPrivate, isTestnet);
    this.ticker = AssetType._getFullTicker(baseTicker, isPrivate);
    this.icon = icon;
    this.numberOfDecimals = numberOfDecimals;
    this.existentialDeposit = existentialDeposit;
    this.isPrivate = isPrivate;
    this.isTestnet = isTestnet;
    this.isNativeToken = isNativeToken;
    this.coingeckoId = coingeckoId;
    this.displayDecimals = displayDecimals;
  }

  static Native(config) {
    if (config.NETWORK_NAME === 'Calamari') {
      return AssetType.Calamari(config, false);
    } else {
      return AssetType.Manta(config, false);
    }
  }

  static Manta(config, isPrivate) {
    return new AssetType(
      getAssetIds(config).MANTA,
      'Manta',
      'MANTA',
      'manta',
      18,
      new BN('100000000000000000'),
      isPrivate,
      'manta-network', // todo: check when manta is added to coingecko
      config.IS_TESTNET,
      true
    );
  }

  static Calamari(config, isPrivate) {
    return new AssetType(
      getAssetIds(config).KMA,
      'Calamari',
      'KMA',
      'calamari',
      12,
      new BN('100000000000'),
      isPrivate,
      'calamari-network',
      config.IS_TESTNET,
      true
    );
  }

  static Acala(config, isPrivate) {
    return new AssetType(
      getAssetIds(config).ACA,
      'Acala',
      'ACA',
      'acala',
      12,
      new BN('100000000000'),
      isPrivate,
      'acala',
      config.IS_TESTNET,
    );
  }

  static Karura(config, isPrivate) {
    return new AssetType(
      getAssetIds(config).KAR,
      'Karura',
      'KAR',
      'kar',
      12,
      new BN('100000000000'),
      isPrivate,
      'karura',
      config.IS_TESTNET,
    );
  }

  static Polkadot(config, isPrivate) {
    return new AssetType(
      getAssetIds(config).DOT,
      'Polkadot',
      'DOT',
      'polkadot',
      10,
      new BN('10000000000'),
      isPrivate,
      'polkadot',
      config.IS_TESTNET,
    );
  }

  static LiquidDot(config, isPrivate) {
    return new AssetType(
      getAssetIds(config).LDOT,
      'Liquid DOT',
      'LDOT',
      'ldot',
      10,
      new BN('500000000'),
      isPrivate,
      'liquid-staking-dot',
      config.IS_TESTNET,
    );
  }

  static Kusama(config, isPrivate) {
    return new AssetType(
      getAssetIds(config).KSM,
      'Kusama',
      'KSM',
      'kusama',
      12,
      new BN('500000000'),
      isPrivate,
      'kusama',
      config.IS_TESTNET,
    );
  }

  static Moonriver(config, isPrivate) {
    return new AssetType(
      getAssetIds(config).MOVR,
      'Moonriver',
      'MOVR',
      'movr',
      18,
      new BN('100000000000000000'),
      isPrivate,
      'moonriver',
      config.IS_TESTNET,
    );
  }


  static Moonbeam(config, isPrivate) {
    return new AssetType(
      getAssetIds(config).GLMR,
      'Moonbeam',
      'GLMR',
      'moonbeam',
      18,
      new BN('1'),
      isPrivate,
      'moonbeam',
      config.IS_TESTNET,
    );
  }

  static Tether(config, isPrivate) {
    return new AssetType(
      getAssetIds(config).USDT,
      'Tether USD',
      'USDT',
      'tether',
      6,
      new BN('1000'),
      isPrivate,
      'tether',
      config.IS_TESTNET,
      false
    );
  }

  static WrappedBitcoin(config, isPrivate) {
    return new AssetType(
      getAssetIds(config).WBTC,
      'Wrapped Bitcoin',
      'WBTC',
      'bitcoin',
      8,
      new BN('35'),
      isPrivate,
      'wrapped-bitcoin',
      config.IS_TESTNET,
      false,
      null,
      8
    );
  }

  static WrappedEthereum(config, isPrivate) {
    return new AssetType(
      getAssetIds(config).WETH,
      'Wrapped Ethereum',
      'WETH',
      'ethereum',
      18,
      new BN('5555555555555'),
      isPrivate,
      'weth',
      config.IS_TESTNET,
      false,
      null,
      8
    );
  }

  static Arbitrum(config, isPrivate) {
    return new AssetType(
      getAssetIds(config).ARB,
      'Arbitrum',
      'ARB',
      'arbitrum',
      18,
      new BN('9000000000000000'),
      isPrivate,
      'arbitrum',
      config.IS_TESTNET,
      false,
      config.NETWORK_NAME === NETWORK.CALAMARI ? 'ARBet': null
    );
  }

  static BinanceCoin(config, isPrivate) {
    return new AssetType(
      getAssetIds(config).BNB,
      'Binance Coin',
      'BNB',
      'bnb',
      18,
      new BN('40000000000000'),
      isPrivate,
      'binance-coin-wormhole',
      config.IS_TESTNET,
      false,
      config.NETWORK_NAME === NETWORK.CALAMARI ? 'BNBet': null,
      4
    );
  }

  static BinanceUsd(config, isPrivate) {
    return new AssetType(
      getAssetIds(config).BUSD,
      'Binance USD',
      'BUSD',
      'busd',
      18,
      new BN('10000000000000000'),
      isPrivate,
      'binance-usd',
      config.IS_TESTNET,
      false,
      config.NETWORK_NAME === NETWORK.CALAMARI ? 'BUSDet' : null
    );
  }

  static Dai(config, isPrivate) {
    return new AssetType(
      getAssetIds(config).DAI,
      'Dai',
      'DAI',
      'dai',
      18,
      new BN('10000000000000000'),
      isPrivate,
      'dai',
      config.IS_TESTNET,
    );
  }

  static UsdCoin(config, isPrivate) {
    return new AssetType(
      getAssetIds(config).USDC,
      'USD Coin',
      'USDC',
      'usdc',
      6,
      new BN('10000'),
      isPrivate,
      'usd-coin',
      config.IS_TESTNET,
      false,
      'USDCet'
    );
  }

  static Polygon(config, isPrivate) {
    return new AssetType(
      getAssetIds(config).MATIC,
      'Polygon',
      'MATIC',
      'polygon',
      18,
      new BN('10000000000000000'),
      isPrivate,
      'polygon',
      config.IS_TESTNET,
      false,
      config.NETWORK_NAME === NETWORK.CALAMARI ? 'MATICet' : null
    );
  }

  static Lido(config, isPrivate) {
    return new AssetType(
      getAssetIds(config).LDO,
      'Lido',
      'LDO',
      'lido',
      18,
      new BN('5000000000000000'),
      isPrivate,
      'lido-dao',
      config.IS_TESTNET,
      false,
    );
  }

  static ShibaInu(config, isPrivate) {
    return new AssetType(
      getAssetIds(config).SHIB,
      'Shiba Inu',
      'SHIB',
      'shibaInu',
      18,
      new BN('1000000000000000000000'),
      isPrivate,
      'shiba-inu',
      config.IS_TESTNET,
      false,
    );
  }

  static Uniswap(config, isPrivate) {
    return new AssetType(
      getAssetIds(config).UNI,
      'Uniswap',
      'UNI',
      'uniswap',
      18,
      new BN('2000000000000000'),
      isPrivate,
      'uniswap',
      config.IS_TESTNET,
      false,
    );
  }

  static Chainlink(config, isPrivate) {
    return new AssetType(
      getAssetIds(config).LINK,
      'Chainlink',
      'LINK',
      'chainlink',
      18,
      new BN('2000000000000000'),
      isPrivate,
      'chainlink',
      config.IS_TESTNET,
      false
    );
  }

  static Apecoin(config, isPrivate) {
    return new AssetType(
      getAssetIds(config).APE,
      'Apecoin',
      'APE',
      'apecoin',
      18,
      new BN('3000000000000000'),
      isPrivate,
      'apecoin',
      config.IS_TESTNET,
      false,
    );
  }


  static AllCurrencies(config, isPrivate) {
    if (config.NETWORK_NAME === NETWORK.CALAMARI) {
      return [
        AssetType.Calamari(config, isPrivate),
        AssetType.Karura(config, isPrivate),
        AssetType.Kusama(config, isPrivate),
        AssetType.Moonriver(config, isPrivate),
        AssetType.Tether(config, isPrivate),
        AssetType.Dai(config, isPrivate),
        AssetType.UsdCoin(config, isPrivate),
        AssetType.WrappedBitcoin(config, isPrivate),
        AssetType.WrappedEthereum(config, isPrivate),
        AssetType.Arbitrum(config, isPrivate),
        AssetType.BinanceCoin(config, isPrivate),
        AssetType.BinanceUsd(config, isPrivate),
        AssetType.Polygon(config, isPrivate),
        AssetType.Lido(config, isPrivate),
        AssetType.ShibaInu(config, isPrivate),
        AssetType.Uniswap(config, isPrivate),
        AssetType.Chainlink(config, isPrivate),
        AssetType.Apecoin(config, isPrivate)
      ];
    } else if (config.NETWORK_NAME === NETWORK.MANTA) {
      return [
        AssetType.Manta(config, isPrivate),
        AssetType.Polkadot(config, isPrivate),
        AssetType.Acala(config, isPrivate),
        AssetType.Moonbeam(config, isPrivate),
        AssetType.LiquidDot(config, isPrivate),
        AssetType.Tether(config, isPrivate),
        AssetType.Dai(config, isPrivate),
        AssetType.UsdCoin(config, isPrivate),
        AssetType.WrappedBitcoin(config, isPrivate),
        AssetType.WrappedEthereum(config, isPrivate),
        AssetType.Apecoin(config, isPrivate)
      ];
    }
  }

  static _getFullName(baseName, isPrivate, isTestnet) {
    let name = isTestnet ? 'Test ' : '';
    if (isPrivate) {
      name += 'zk';
    }
    return name + baseName;
  }

  static _getFullTicker(baseTicker, isPrivate) {
    return isPrivate ? `zk${baseTicker}` : baseTicker;
  }

  toPrivate() {
    return new AssetType(
      this.assetId,
      this.baseName,
      this.baseTicker,
      this.icon,
      this.numberOfDecimals,
      this.existentialDeposit,
      true,
      this.coingeckoId,
      this.isTestnet,
      this.isNativeToken,
      this.logicalTicker,
      this.displayDecimals
    );
  }

  toPublic() {
    return new AssetType(
      this.assetId,
      this.baseName,
      this.baseTicker,
      this.icon,
      this.numberOfDecimals,
      this.existentialDeposit,
      false,
      this.coingeckoId,
      this.isTestnet,
      this.isNativeToken,
      this.logicalTicker,
      this.displayDecimals
    );
  }

  toggleIsPrivate() {
    if (this.isPrivate) {
      return this.toPublic();
    } else {
      return this.toPrivate();
    }
  }

  canTransferXcm = (originChain, destinationChain) => {
    return (
      originChain.xcmAssets.find((asset) => asset.name === this.name) &&
      destinationChain.xcmAssets.find((asset) => asset.name === this.name)
    );
  };
}
